services/src/main/java/io/grpc/services/BinlogHelper.java

```diff
@@ -332,56 +387,86 @@ private static Deadline min(@Nullable Deadline deadline0, @Nullable Deadline dea
     return deadline0.minimum(deadline1);
   }
 
-  public ClientInterceptor getClientInterceptor(final CallId callId) {
+  interface TimeProvider {
+    /** Returns the current nano time. */
+    long currentTimeNanos();
+
+    TimeProvider SYSTEM_TIME_PROVIDER = new TimeProvider() {
+      final long offsetNanos =
+          TimeUnit.MILLISECONDS.toNanos(System.currentTimeMillis()) - System.nanoTime();
```

Comment:
If the clock changes, then the timestamps will be incorrect. Is that what is supposed to happen?